### PR TITLE
Roles support added

### PR DESCRIPTION
**Pull request by Exsilium122**
I have added support for roles. Unfortunately not everything is backward compatible. If you are going to accept the pull request I can also change your sample app to be compatible with this patch.

Samples how to use new code:

    div#navbar-collapse.navbar-collapse.collapse
        ul.nav.navbar-nav
            li(repeat.for="row of router.navigation | authFilter:auth.tokenPayload", class="${row.isActive ? 'active' : ''}")
                a(data-toggle="collapse", data-target="#navbar-collapse.in", href.bind="row.href")
                    span.fa.fa-fw(class="fa-${$index + 1}")
                    span ${row.title}
            ul.nav.navbar-nav.navbar-right(if.bind="!auth.withRoles().isAuthenticated")
                li
                    a(click.trigger="login()", href="#login")
                        span.fa.fa-fw(class="fa-100")
                        span ${'router.login' | t}
                li
                    a(click.trigger="signUp()", href="#signup")
                        span.fa.fa-fw(class="fa-101")
                        span ${'router.signUp' | t}
            ul.nav.navbar-nav.navbar-right(if.bind="auth.withRoles().isAuthenticated")
                li
                    a(click.trigger="logout()", href="#logout")
                        span.fa.fa-fw(class="fa-101")
                        span ${'router.logout' | t}
        span.navbar-throbber.fa.fa-refresh.fa-spin.fa-2x(if.bind="router.isNavigating")

Configuration of router is backword compatible. _auth_ can be boolean as before, and the string array to specify which roles are required. Empty array means just authenticated. 

    config.map(
        [
            { route: ['', 'lightingSystem', 'lightingSystem/:oid'], name: 'lightingSystem', moduleId: 'pages/lightingSystem', /*href: '#lightingSystem',*/ nav: true, title: this.i18n.tr('router.lightingSystem')},
            { route: ['libraries', 'libraries/:type', 'libraries/:type/:key'], name: 'libraries', moduleId: 'pages/libraries', nav: true, title: this.i18n.tr('router.libraries')},
            { route: 'admin', name: 'admin', moduleId: 'pages/admin', nav: true, title: this.i18n.tr('router.admin'), auth: ['admin']}, 
            { route: 'users', name: 'users', moduleId: 'pages/users', nav: true, title: this.i18n.tr('router.users'), auth: ['admin', 'teamlead']},
            { route: 'profile', name: 'profile', moduleId: 'pages/profile', nav: true, title: this.i18n.tr('router.profile'), auth: true},
            { route: 'login', name: 'login', moduleId: 'pages/login', title: this.i18n.tr('router.login')}
        ]
    );

### CODE REVIEW

**Comment by paulvanbladel** [comment]
Hi Alex,
Thanks a lot.
To be honest and hope you don't mind, I'm not a big fan of roles as a means for controlling authorization. I prefer the more modern 'claims' approach because it's more action based. 
When adding roles also to the routing config, it all becomes quite heavy. In my view, this type of authorization logic should be better handled via following viewmodel life cyle hook point: 

canActivate(params, routeConfig, navigationInstruction) 

By doing so one can add real code based logic as well. 

Anyway, I'll study more in depth your approach.

**Comment by Exsilium122** [comment]
Hi,

to be honest it is first time I read something about claims authorization. However from short reading my solution is not so far away from it.

In my solution it is a server who create JWT (token? do I need to add it as T already stands for token) which contains payload. In the payload I added a field _roles_ which is an array of roles that the user has. So it might be understood as:
I claim I am:
- admin
- user
- accountant

Then aurelia application uses it to allow access to certain parts of it and functionality. 
Configuration roles in router it is just one possibility. Using this approach no coding is required and routes will be visible and accessible only for users with given roles.
Of course one might use manual coding to achieve similar functionality for widgets. So configuration in router it is just shortcut. 

Probably I messed something up so please show me the error in my understanding of "claims based authorization"

**Comment by RWOverdijk** [comment]
This has become quite a messy PR. I get what you're going for, though. I've poked you on gitter about this, I'm working on a plugin that will work _with_ aurelia-auth. There's someone else on it, too. So the three of us could surely put together something nice.

https://gitter.im/RWOverdijk

**Comment by paulvanbladel** [comment]
@Exsilium122 
@RWOverdijk  thanks for joining, we can discuss things here as well, since we are talking about aurelia-auth, no need for private discussions.
Hi Alex,
Sorry for the confusion. What I mean is that roles are basically a 'container' and it is a better idea not to link code to roles, but use rather a more fine grained concept, something like permissions.
What i did not mean is that these permissions should be part of the JWT, they can but it's not practical I think.
Roles can then still be used to group a series a permissions.
E.g. role admin has permission:
CanHardDeleteCustomer, CanEditAllCustomer, ..
Role agent has permission: CanSoftDeleteCustomer, CanEditOwnCustomer, ..
So, the link between the asset that you want to protect (e.g. show screen, show button) should be linked to the permission rather than the role.
What we need boils down to:
1. a global state object for carrying the current user and the set of permissions
2. a mechanism to retrieve the set of permissions from the backend (or the STS). Since this is heavily backend dependent, we need flexibility here, e.g. provider model. We definitely want to share the set of permissions between front end and backend. At the backend the permissions are used for securing real assets, while at the front-end we simply provide 'security related comfort' (e.g. hiding things). As you know, a SPA is by definition insecure.
3. a mechanism for annotating which permission is linked to which asset. In the .net world, we often use  attributes for this. Maybe we can use ES7 decorators ?

cheers
paul.

**Comment by Exsilium122** [comment]
**Ad 1.** So it can be JWT, but can be as well a token received independantly from 3rd party STS. 
**Ad 2.** If it is a part of JWT then backend is aware of if authenticy as it was this backend which signed it. 
If external STS is used then our backend when suddenly receive such token should check if validity. If it was signed by STS or not. Technically it can be done with asymmetric cryptography. 
**Ad 3.** Sound interesting. Especially annotations ES7. How this can be acheived in Aurelia? I mean the lifecycle of widgets. For routes it would be easy. We will not display the route if permissions are not met, instead redirect to error page or login page. 
What about widgets? Is it possible not to compose the widget into the page dynamically. Of course I can add if.bind statement manually. But is it possible to make it more automatic? So the developer compose the widget into view. The widget is annotated with @permission('CanListSecretFiles'). Then it will be visible on dashboard only for right users?

Is there any opinionated standard for STS? Wouldn't it be easier to first make a support for JWT based delivery. Then extend it to STS?

**Comment by wshayes** [comment]
I started off quite excited about storing things like permissions and other user profile data in the JWT, but then I realized that the token would have to somehow be expired and re-issued if the user updated any of that information in order to be able to use it.  I've since only planned on putting immutable information in the JWT payload (and permissions or roles is definitely not immutable).

**Comment by Exsilium122** [comment]
@wshayes Everything client site is "mutable". You should not be worry about JWT integrity. Yes client site can modify it. Yes it will show restricted areas on client side. BUT: Server do not use trust client site. Integrity of token if verified. All content f token is signed. Therefore any manipulation of token by user will result in server refusal of doing anything. So user will see restricted widget, but those widgets will not be populated with any data. Moreover even without token manipulation user can read javascript code and recreate those widgets (hard and boring but dooable)

Conclusion. It is safe to store any kind of information in token, unless it is a secret. Of course size also meters as token will be probably sent with every request to server.

**Comment by wshayes** [comment]
Sorry I wasn’t clear.  I agree - don’t trust anything client side.  My initial thoughts around JWT is that it was a good ‘state’ variable to hold user profile information on the client side.  BUT - it is hard to expire and update if the user changes their name or gets new permissions/claims.  If your client is using the JWT payload to hold this mutable (on the server side mutable) information, your client won’t be updated in a timely manner - it will have to wait for a new JWT to be issued.  Kind of mean to require the user to logout and back in again to be able to use their new permissions.  I know the new permissions really matter on the server side, but I would generally only allow users to see functionality that they are permitted (not that they can’t hack the front end code and do whatever they want on the front end - they just would be able to complete the transaction on the server.

Bottom line - I only include truly user immutable content in the JWT payload that I would need on the front end which for one application I am working on is their userid since I allow users to update their email address.

**Comment by Exsilium122** [comment]
@wshayes so why not fetch refreshed token every time the application bootstrap?

Refreshing the token in every request will be not optimal, or even pointless. It will break the idea of JWT, because backend will have to authorize user every time with database or keep server session. With JWT you can use less resources server side.

The above bad solution you can achieve with what you described. Keep only constant thing in token. But then every time "refresh" permissions from server, which have the same consequences as previous paragraph. 

So I think that refreshing the token, at bootstrap is optimal. By refresh I mean update permissions but leave expire date the same. So user will have to provide credentials from time to time.

**Comment by wshayes** [comment]
I think we are agreeing mostly.  I also don’t believe that the token should be updated on every server request.  But if my user gets a new permission and the token won’t expire for several hours, how would the user be able to utilize their new permission functionality before the token gets refreshed?

**Comment by Exsilium122** [comment]
F5 application.

Or make interval check, every 1 minute to update the token ;)

**Comment by Vidarls** [comment]
@Exsilium122 @wshayes Access tokens from an STS are ment to give authorization to high level resources (Scopes), an API / a set of APIs etc. Anything more application specific should be handled by application (server side or client side). If not you will be coupling your STS very tightly to your business domain. To get additional info, you use your access token to query your backend server for application specific capabilities.

Tokens may have very short or very long expiry times, and the developer of a client may not be in a position to decide long or short. What are the consequences of telling your users to log in again when they get a new permission? Will something very bad happen if there are some delay before they receive it?
Most STS will not even force a new password entry when a user is redirected to login page, as they often have long lived cookies authenticating them, meaning a new "login" from the clients perspective will just be a couple of redirects, no user interactions required. (You could even have a "refresh permissions" button that fetches a new token using existing non-expired token without redirects.)

**Comment by Exsilium122** [comment]
@Vidarls Thanks for some explanation.

I was not focusing on STS. I was writing about or own JWT issued by our server with fine grained application specific permission set. Anyway once more thanks for some more info about STS.

**Comment by paulvanbladel** [comment]
@Vidarls : couldn't agree more.
@Exsilium122 : basically the security involved for transferring permissions from server to client is completely of the same magnitude than transferring let's say customers. Anyhow, the SPA is an insecure thing, a malicious user can always bypass permissions and access screens and other artefacts, even if you package permissions in tokens with nec plus ultra security pepper and salt. The real security happens on the API level: we can not prevent that a malicious user access the 'customer' screen, but we can prevent that he can do crud operations on the customer api server side. Bottomline, the backend of the app should simply expose an endpoint where an authenticated user can retrieve his/her permissions.
